### DISCUSSION TITLE
UUID Primary Key

### DISCUSSION THREAD

**Discussion by mohux** [adonisjs/core#1296] [Help]
I'm using AdonisJS v5 and not v4
when i use User.create()

and add @beforeCreate hook to create a uuid 
it returns 0 instead of the stored uuid

therefore the solution of adding 
```static async get increments(){}```
does not work

I tried so many things to fix the issue
but it always return 0

**Comment by RomainLanz** [comment]
Hey @mohux! 👋 

There's no `increments` method inside the Lucid Model.
Also, if you are using PostgreSQL I'd recommend you to directly use the function `uuid_generate_v4()` as default instead of generating the `uuid` yourself.

**Reply by mohux** [reply]
am using mySQL, and in V5 the uuid is stored correctly in the DB, its just that the model returns 0 when I save the model
if I rename "id" to anything else and keep it as primary, it works for some reason

**Comment by fnoquiq** [comment]
This post worked for me

https://forum.adonisjs.com/t/how-to-use-uuid-as-primary-key-postgre/3918

**Reply by mohux** [reply]
Hello, thanks for replaying, this didnt work for me on adonis5 

the issue is weird, it happens when i use my primary key as uuid and name it "id"

if i name it "user_id" it returns the stored one
btw even when it returns 0, it stores the uuid correctly in the database
its just the returned value

**Comment by mohux** [comment]
I figured out that the problem is of how Lucid treats primaryKey via @column({isPrimary:true})
it treats it as increments by default even if your migration is uuid
still cant find a way to solve it,

**Reply by RomainLanz** [reply]
We cannot help you more with the amount of information you shared.
I'm using UUID in all of my project with no change to my model.

**Reply by mohux** [reply]
If your using V5

and write this

![image](https://user-images.githubusercontent.com/57658626/88151051-dbf18800-cc0a-11ea-93e7-38d5f641bf73.png)

and inside the model I do the following 
![image](https://user-images.githubusercontent.com/57658626/88151109-f166b200-cc0a-11ea-9ba8-e98029d2d145.png)

then i create a user with create method
![image](https://user-images.githubusercontent.com/57658626/88151166-02afbe80-cc0b-11ea-8cd4-1ee82321de18.png)


the returned user has id = 0

i was able to solve it by doing so
![image](https://user-images.githubusercontent.com/57658626/88151213-15c28e80-cc0b-11ea-8e96-4872b15130db.png)

but not sure if its the correct way

**Reply by thetutlage** [reply]
Lemme check with Mysql

**Reply by mohux** [reply]
if I use my crazy solution
the find method won't return the ID 😢

**Reply by mohux** [reply]
it works perfectly now 
![image](https://user-images.githubusercontent.com/57658626/88153743-7bfce080-cc0e-11ea-82de-77fc90a64712.png) 
but pretty sure it isn't the meant solution

**Comment by mohammadou1** [comment]
Did anyone find a solution? i tried it with PostgreSQL, its working so perfectly, without the need to use consume function

**Comment by dajpes** [comment]
I'm also using Mysql(MariaDb), and I use the beforeCreate hook in order to add an uuid:


``` Javascript

import { v4 as uuidv4 } from "uuid";

export default class User extends BaseModel {

 @beforeCreate()
  public static async addUidHook(user: User) {
    user.uid = uuidv4();
  }

...

  @column({ isPrimary: true, columnName: "uid" })
  public uid: string;
}


```

**Reply by BenCherif** [reply]
@mohux  i've used refresh() function after create the object  " await user.refresh()"  then get the id correctly
@dajpes  your answer is working for me thank u .

**Reply by solnsumei** [reply]
> @mohux i've user refresh() function after create the object " await user.refresh()" then get the id correctly
> @dajpes your answer is working for me thank u .

If you refresh, adonis returns a random data

**Reply by BenCherif** [reply]
> > @mohux i've user refresh() function after create the object " await user.refresh()" then get the id correctly
> > @dajpes your answer is working for me thank u .
> 
> If you refresh, adonis returns a random data

you are right ,so i generate a random uuidv4() and i put it to the object so when object has been created i fetch it by id that i've generated before creation . i didn't find a better way :/

**Comment by juni0r** [comment]
If you use Postgres, there's a clean DRY way to add UUID primary keys with auto-generated IDs (no `@beforeCreate`)

#### 1. Enable the UUID extension (migration must precede any models using UUIDs)

```javascript
import BaseSchema from '@ioc:Adonis/Lucid/Schema'

export default class extends BaseSchema {
  public up() {
    this.schema.raw('CREATE EXTENSION IF NOT EXISTS "uuid-ossp"')
  }

  public down() {
    this.schema.raw('DROP EXTENSION IF EXISTS "uuid-ossp"')
  }
}
```

#### 2. Create migrations for your models

```javascript
import BaseSchema from '@ioc:Adonis/Lucid/Schema'

export default class extends BaseSchema {
  protected tableName = 'my_model'

  public async up() {
    this.schema.createTable(this.tableName, (table) => {
     // use Postgres builtin function 'uuid_generate_v4' to auto-generate an id
      table.uuid('id').primary().defaultTo(this.raw('uuid_generate_v4()'))

     // more columns…
    })
  }

  // …
}
```

#### 3. Define your model class

```javascript
export default class MyModel extends BaseModel {
  @column({ isPrimary: true })
  public id: string

  // ...
}
```

Refer to primary uuid columns in related models:

```javascript
export default class extends BaseSchema {
  public async up() {
    this.schema.createTable('my_related_model', (table) => {
      // …
      table.uuid('my_model_id').references('id').inTable('my_model').onDelete('CASCADE')
      // …
    })
  }
}
```

**Reply by McSneaky** [reply]
One additional thing I'd suggest to do, is to add `uuid-ossp` to `pg_catalog`

It makes dealing with backups, restores and clones lil-bit easier. Otherwise you might end up in situations where you want to clone database within same server, but suddenly uuid extension is not installed, since it was only added to specific database. Adding it to pg_catalog adds it to future clones too (and new databases, have to keep that in mind, you might not want it enabled in new DBs for some reason)

```ts
this.schema.raw('CREATE EXTENSION IF NOT EXISTS "uuid-ossp"')
```
Becomes
```ts
this.schema.raw('CREATE EXTENSION IF NOT EXISTS "uuid-ossp" schema pg_catalog version "1.1";')
```

**Comment by Forthtilliath** [comment]
Here my solution, since previews code didn't worked for me on adonis 6.

Insert uuid from model
```ts
export default class User extends compose(BaseModel, AuthFinder) {
  static selfAssignPrimaryKey = true

  @column({ isPrimary: true })
  declare id: string

  @beforeCreate()
  static async setId(user: User) {
    user.id = crypto.randomUUID()
   }
}
```

~~Solution 2 : insert uuid from schema~~ Not working !
```ts
this.schema.createTable(this.tableName, (table) => {
   table.uuid('id').primary().defaultTo(crypto.randomUUID())
})
```

**Reply by RomainLanz** [reply]
```ts
this.schema.createTable(this.tableName, (table) => {
   table.uuid('id').primary().defaultTo(crypto.randomUUID())
})
```

This code will not work. You have to define the PSQL function to use it. In your example, you are defining a static value computed when the migration runs to be your default value.

**Reply by Forthtilliath** [reply]
You're alright. I checked after only one new user...

**Comment by vitorabner** [comment]
Another solution using preload:

```
node ace make:preload hooks
```

```
hooks.ts

import { BaseModel } from '@adonisjs/lucid/orm'

const boot = BaseModel.boot

BaseModel.boot = function () {
  if (this.booted) return

  boot.call(this)

  this.before('create', (item) => {
    item.$setAttribute('id', crypto.randomUUID())
  })
}
```

This code will run for every model.

**Reply by adityadarma** [reply]
I have try this, but get error at User model

error: insert into "user_with_user_finders" ("created_at", "email", "id", "name", "password", "updated_at") values ........


then other model also is booted
